src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/PresentationHandlerBase.cs

```diff
@@ -0,0 +1,100 @@
+
+        protected abstract string MethodName { get; }
+
+        protected PresentationHandlerBase(
+            LSPRequestInvoker requestInvoker,
+            LSPDocumentManager documentManager,
+            LSPProjectionProvider projectionProvider,
+            LSPDocumentMappingProvider documentMappingProvider,
+            HTMLCSharpLanguageServerLogHubLoggerProvider loggerProvider)
+        {
+            _requestInvoker = requestInvoker;
+            _documentManager = documentManager;
+            _projectionProvider = projectionProvider;
+            _documentMappingProvider = documentMappingProvider;
+
+            _logger = loggerProvider.CreateLogger(nameof(PresentationHandlerBase<TParams>));
+        }
+
+        protected abstract TextDocumentIdentifier GetTextDocumentIdentifier(TParams request);
+
+        protected abstract Range GetRange(TParams request);
+
+        public Task<WorkspaceEdit?> HandleRequestAsync(TParams request, ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            var textDocument = GetTextDocumentIdentifier(request);
+            var range = GetRange(request);
+
+            return HandleRequestAsync(request, textDocument, range, cancellationToken);
+        }
+
+        private async Task<WorkspaceEdit?> HandleRequestAsync(TParams request, TextDocumentIdentifier textDocumentIdentifier, Range range, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Starting request for {textDocumentIdentifier.Uri}.");
+
+            if (!_documentManager.TryGetDocument(textDocumentIdentifier.Uri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            if (!documentSnapshot.TryGetVirtualDocument<HtmlVirtualDocumentSnapshot>(out var htmlDoc))
+            {
+                _logger.LogWarning($"Failed to find virtual HTML document for {textDocumentIdentifier.Uri}.");
+                return null;
+            }
+
+            var projectionResult = await _projectionProvider.GetProjectionAsync(documentSnapshot, range.Start, cancellationToken).ConfigureAwait(false);
+            if (projectionResult is null)
+            {
+                return null;
+            }
+
+            if (projectionResult.LanguageKind is not (RazorLanguageKind.Html or RazorLanguageKind.CSharp))
+            {
+                _logger.LogInformation($"Unsupported language {projectionResult.LanguageKind:G}.");
+                return null;
+            }
+
+            var languageServerName = projectionResult.LanguageKind == RazorLanguageKind.CSharp ? RazorLSPConstants.RazorCSharpLanguageServerName : RazorLSPConstants.HtmlLanguageServerName;
+
+            var textBuffer = htmlDoc.Snapshot.TextBuffer;
+            var result = await _requestInvoker.ReinvokeRequestOnServerAsync<TParams, WorkspaceEdit?>(
+                textBuffer,
+                MethodName,
+                languageServerName,
```

Comment:
>languageServerName, [](http://example.com/codeflow?start=16&length=19)

out of curiosity, isn't there a version of this method that doesn't take in the language server name?

Reply:
Nope, they all take language server name

Reply:
Yeah, I guess I'm not sure how tightly coupled you guys like being with our language server, but there is a ReinvokeRequestOnServerAsync method that doesn't take in the language server name